Switching cloud manifest during leader election

seckcoder opened the pull request:
Added three functions to support switching cloud manifest during leader election.
1. `ApplyLocalCloudManifestDelta` will update local CM/M based on the delta changes.
2. `UploadLocalCloudManifest` will upload local CM/M to s3.
3. `NewManifestOnNextUpdate` will make next manifest update to reset `descriptor_log_` and write latest snapshot of VersionSet to a new MANIFEST file

On leader, during leader election, we will (assuming state=RUNNING):
```
new_epoch = tryBecomeLeader()
ApplyLocalCloudManifestDelta(new_cookie)
CAS new_cookie
UploadLocalCloudManifest()
NewManifestOnNextUpdate()
Generate kNewEpoch with the delta
```

On follower, once it gets the `kNewEpoch` event, it will
```
ApplyLocalCloudManifestDelta(new_cookie)
NewManifestOnNextUpdate()
```

seckcoder commented on cloud/cloud_env_impl.cc:
```diff
@@ -1840,6 +1844,150 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
+
+  // rolling epoch of the Manifest file
+  auto new_epoch = generateNewEpochId();
+  cloud_manifest->AddEpoch(max_file_num, new_epoch);
```
@igorcanadi Can we assume that there is only one single thread calling `Roll/Switch CloudManifest` for one shard?

It can be an issue when it's called in multi-threads since `cloud_manifest` can point to different content here (compared to calling `GetCurrentEpoch` above).

Or we should move thread safety protection for cloud manifest into `CloudEnv` by protecting the pointer.

igorcanadi replied:
Yes, we should add this to the contract of the API. It shouldn't be the problem to guarantee that on the caller.

igorcanadi commented on cloud/cloud_env_impl.h:
```diff
@@ -366,6 +407,9 @@ class CloudEnvImpl : public CloudEnv {
   std::chrono::seconds file_deletion_delay_ = std::chrono::hours(1);
   std::unordered_map<std::string, int> files_to_delete_;
+
+  // mutex to protect updates against cloud_env_options
+  port::RWMutex cloud_env_option_mutex_;
```
Introducing this mutex means we need to lock `cloud_env_option_mutex_` in read mode for every access to `cloud_env_options`, which is definitely not the case today.

I don't think we need to mutate cookie ever, though. See my other comments.

seckcoder replied:
Makes sense. In that case, we need to rename `cookie` to `cookie_when_open` as well, since it might point to switched CloudManifestFile.

igorcanadi commented on cloud/cloud_env_impl.cc:
```diff
@@ -1840,6 +1844,150 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
+  return st;
+}
+
+Status CloudEnvImpl::SwitchCloudManifest(const std::string& local_dbname,
+                                         const std::string& new_cookie) {
+  std::string current_cookie = GetCookie();
+  if (new_cookie == current_cookie) {
+    return Status::InvalidArgument(
+        "New cloud manifest to switch to should have different cookie");
+  }
+
+  auto st = FetchCloudManifest(local_dbname, new_cookie);
+  if (!st.ok()) return st;
+
+  auto new_cloud_manifest_file_name =
+      MakeCloudManifestFile(local_dbname, new_cookie);
+  std::unique_ptr<SequentialFileReader> log;
+  st = SequentialFileReader::Create(GetBaseEnv()->GetFileSystem(),
+                                    new_cloud_manifest_file_name, FileOptions(),
+                                    &log, nullptr);
+  if (!st.ok()) return st;
+
+  st = cloud_manifest_->ReloadFromLog(std::move(log));
+  if (!st.ok()) return st;
+
+  SetCookie(new_cookie);
+  return st;
+}
```
I don't think we should ever need to switch cloud manifests by downloading them from S3 and reloading the state from S3.

How about this:
* On startup, download the CLOUDMANIFEST with the cookie provided.
* If you are the leader, roll the cloud manifest by writing the delta to the local CLOUDMANIFEST file, write the delta to the replication log and also upload the CLOUDMANIFEST to S3 with the new cookie.
* If you are the follower, apply the delta you get from the replication log to the local CLOUDMANIFEST

 You only need to provide two new functions:
 1. RollCloudManifest(delta)
 2. UploadCloudManifestToCloud(new_cookie)
 
 Here delta is (max_file_number, epoch).
 
You call (1) on both the leader and the follower. (2) is only applicable to the leader.

seckcoder replied:
Also, are we going to call `RollCloudManifest(delta)` before or after CAS(cookie)?

I think either way has some issues:

- If we `CAS(cookie)` before `RollCloudManifest(delta)`, then the cloud manifest file is not ready in s3 yet after `CAS(cookie)`
- If we `CAS(cookie)` after `RollCloudManifest(delta)`, it's possible that `CAS(cookie)` fails, but the `CloudManifest` content has been changed on this follower. In the future, this follower will always contain an extra `pastEpoch`

I'm thinking about this:
```
// On Leader
RollCloudManifest(new_cookie, &delta) // fetch latest manifest file, upload CloudManifest and new Manifest to s3, return delta as serialized string
CAS(cookie)
ApplyCloudManifestDelta(delta) // apply delta to cloud manifest
Replicate delta to followers

// On Follower
ApplyCloudManifestDelta(delta) // apply delta to cloud manifest
```

igorcanadi replied:
> But shouldn't we need to download the latest MANIFEST-epoch from s3 (on leader) to figure out the max_file_number? The local manifest file could be different from the one storing in s3. And I believe it might be better to do this in CloudEnv.

We will only roll the cloud manifest after opening the database, right? The process of opening the database itself will download the latest MANIFEST-epoch and read it. We will never need to re-download MANIFEST on a running database.

> Also, are we going to call RollCloudManifest(delta) before or after CAS(cookie)?

We need to do Roll before CAS, because we need CLOUDMANIFEST-cookie to exist before switching the cookie, as you point out.

> If we CAS(cookie) after RollCloudManifest(delta), it's possible that CAS(cookie) fails, but the CloudManifest content has been changed on this follower. In the future, this follower will always contain an extra pastEpoch. This can be a big issue, since on this follower, the same file number can be mapped to multiple epochs in CloudManifest.

There are two scenarios when CAS(cookie) fails:
1/ The new leader recovered the NewEpochRecord that contains the delta to the CLOUDMANIFEST from the old leader. In that case the leader and the follower should have exactly the same CLOUDMANIFEST contents.
2/ The new leader didn't recover the NewEpochRecord that the follower has applied to the local CLOUDMANIFEST. This is a diverged branch and the follower will have to resync from the new leader.

seckcoder replied:
> We will only roll the cloud manifest after opening the database, right? 
Wouldn't the follower need to roll cloud manifest whenever it tries to become leader? (and we want to avoid reopening the db here). 

> There are two scenarios when CAS(cookie) fails:
I don't fully get this part. IMO, CAS(cookie) can fail if there are two followers trying to become leader simultaneously. Only one of them should succeed when CAS(cookie)

seckcoder commented:
Updated, PTAL

igorcanadi commented on db/db_impl/replication_codec.cc:
```diff
@@ -4,15 +4,18 @@
   if (!GetFixed16(src, &version)) {
-    return Status::Corruption("Unable to decode memtable switch record version");
+    return Status::Corruption(
+        "Unable to decode memtable switch record version");
```
Don't we also need to verify that `version == 1` here?

igorcanadi commented on db/db_impl/replication_codec.cc:
```diff
@@ -4,15 +4,18 @@
-  PutFixed16(dst, 1); // version
+Status SerializeMemTableSwitchRecord(std::string* dst,
+                                     const MemTableSwitchRecord& record) {
+  PutFixed16(dst, 1);  // version
```
`PutFixed8` should suffice here.

igorcanadi commented on cloud/manifest_reader.h:
```diff
@@ -26,12 +26,20 @@ class ManifestReader {
+  // constructed cloud_manifest
   Status GetLiveFilesAndCloudManifest(
-      const std::string bucket_path, std::set<uint64_t>* list,
+      const std::string bucket_path,
+      const std::string cookie,
```
Passing argument by const value is weird. Are you sure this shouldn't be `const std::string&`?

igorcanadi commented on cloud/manifest_reader.h:
```diff
@@ -26,12 +26,20 @@ class ManifestReader {
+    // reopening the shard. We should fix this by reading from
+    // cenv_->cloud_manifest directly
+    return GetLiveFilesAndCloudManifest(
+        std::move(bucket_path), cenv_->GetCloudEnvOptions().cookie_on_open,
```
`std::move(bucket_path)` doesn't move because `bucket_path` is const. See below, too.

igorcanadi commented on cloud/cloud_manifest.cc:
```diff
@@ -108,9 +102,25 @@ Status CloudManifest::LoadFromLog(std::unique_ptr<SequentialFileReader> log,
+
+}  // namespace
+
+// Format:
+// header: format_version (varint) number of records (varint)
+// record: tag (varint, 1 or 2)
+// record 1: epoch (slice), file number
+// record 2: current epoch
```
Move this above `LoadEpochsFromLog`?

igorcanadi commented on cloud/cloud_manifest.cc:
```diff
@@ -108,9 +102,25 @@ Status CloudManifest::LoadFromLog(std::unique_ptr<SequentialFileReader> log,
+                                  std::unique_ptr<CloudManifest>* manifest) {
+  std::string currentEpoch;                                
+  std::vector<std::pair<uint64_t, std::string>> pastEpochs;
+  auto st = LoadEpochsFromLog(std::move(log), currentEpoch, pastEpochs);
```
Why extract this in the separate function? As far as I can tell the behavior is exactly the same.

seckcoder replied:
oops. It's introduced in previous commit. I should remove it

igorcanadi commented on db/db_impl/replication_codec.cc:
```diff
@@ -32,4 +35,27 @@ std::string RecordMemTableSwitch(
 }
+
+Status SerializeCloudManifestDelta(std::string* dst, CloudManifestDelta delta) {
+  PutVarint64(dst, delta.file_num);
```
Maybe add a version to this, just like `MemtableSwitch` above?

Edit: Although, see above, we might not need to serialize the cloud manifest delta in RocksDB codebase.

seckcoder replied:
actually, I'm thinking about removing the version in `MemtableSwitch`. Seems it's not necessary. If we introduce some backwards incompatible change, we can always introduce a new event type(e.g., `kMemtableSwitch2` )

igorcanadi commented on cloud/cloud_env_impl.h:
```diff
@@ -266,6 +273,28 @@ class CloudEnvImpl : public CloudEnv {
 
+  std::string CloudManifestFile(const std::string& dbname);
+
+  // Roll the cloud manifest happens when follower tries to become the leader.
+  // It will:
+  //
+  // - Generate new delta based on the specified next_file_num, which represents
+  // the snapshot we cut for the new cloud manifest. SST files with filenum >=
+  // next_file_num belongs to the new epoch
+  // - Apply the delta locally
+  // - Upload CLOUDMANIFEST and MANIFEST to s3
+  //
+  // `cloud_manifest_delta` contains the generated cloud manifest delta during rolling.
+  //
+  // NOTE: If any step after(include itself) this, and before we write
+  // `kNewEpoch` with the delta, fails, we would have to reopen the db. Reason is the local cloudmanifest
+  // contains dirty delta generated during rolling and we have to reopen db to clean it up.
+  Status RollCloudManifest(std::string local_dbname, std::string new_cookie,
+                           uint64_t next_file_num,
+                           std::string* cloud_manifest_delta);
+  // Apply the serialized cloud manifest delta locally.
+  Status ApplyCloudManifestDelta(std::string cloud_manifest_delta);
```
These two functions are not really orthogonal. How about this:

```
Status ApplyLocalCloudManifestDelta(const std::string& dbname, uint64_t next_file_num, std::string new_epoch);
Status UploadLocalCloudManifest(const std::string& dbname, std::string& new_cookie);
```

This way you run `ApplyLocalCloudManifestDelta()` in two situations -- applying remote cloud manifest delta and generating the cloud manifest delta yourself.

Leader runs `UploadLocalCloudManifest()` to upload its local CM and M to the cloud.

Feel free to let the upper layers generate the new cloud manifest epoch. We can do that in the leaf codebase.

Same with serializing the cloud manifest delta, we can do that in the leaf codebase.

BTW currently our local MANIFEST filename is `MANIFEST-epoch`. I wonder if this should be just `MANIFEST` (note that cloud filename would still be `MANIFEST-epoch`). That way you don't need to rename the manifest file when you roll the new epoch in cloud manifest. Renaming the manifest file is tricky on a running database.

seckcoder replied:
why do you think it's tricky? While renaming the file here, there shouldn't be running threads accessing it right?

igorcanadi replied:
Yeah, but if we can easily avoid having that constraint we should.

seckcoder replied:
If we don't have the constraint, wouldn't we have similar issue even if writing to the same manifest file?

There could be two threads updating manifest file at the same time and we would need lock to coordinate them.

igorcanadi commented on cloud/cloud_env_impl.cc:
```diff
@@ -1783,7 +1789,7 @@ Status CloudEnvImpl::CreateCloudManifest(const std::string& local_dbname) {
 Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
   assert(cloud_env_options.roll_cloud_manifest_on_open);
-  auto oldEpoch = GetCloudManifest()->GetCurrentEpoch().ToString();
+  auto oldEpoch = GetCloudManifest()->GetCurrentEpoch();
```
Could we build `RollNewEpoch` on top of our newly-introduced functions?

igorcanadi commented on cloud/cloud_env_impl.cc:
```diff
@@ -1840,6 +1843,45 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
+    return st;
+  }
+
+  // upload the cloud manifest file corresponds to cookie (i.e., CLOUDMANIFEST-cookie)
+  return GetStorageProvider()->PutCloudObject(
+      MakeCloudManifestFile(local_dbname, cookie), GetDestBucketName(),
+      MakeCloudManifestFile(GetDestObjectPath(), cookie));
```
Could you also upload the cloudmanifest file in the old format (`CLOUDMANIFEST`) even when the cookie is set? That way we can quickly roll back to the world without leader follower, which doesn't have the cookie and expects `CLOUDMANIFEST` to be there.

igorcanadi commented on cloud/cloud_env_impl.cc:
```diff
@@ -851,15 +851,15 @@ Status CloudEnvImpl::LoadLocalCloudManifest(const std::string& dbname) {
     std::unique_ptr<CloudManifest>* cloud_manifest) {
   std::unique_ptr<SequentialFileReader> reader;
-  auto cloud_manifest_file_name = CloudManifestFile(dbname);
+  auto cloud_manifest_file_name = MakeCloudManifestFile(dbname, cookie);
```
See below. I'm thinking our local cloudmanifest filename could be without the cookie suffix.

igorcanadi commented on cloud/db_cloud_test.cc:
```diff
@@ -1800,6 +1809,103 @@ TEST_F(CloudTest, FileCacheOnDemand) {
+  ASSERT_OK(db_->Put(WriteOptions(), "Hello", "new_world"));
+  ASSERT_OK(db_->Flush(FlushOptions()));
+
+  // reopen with cookie = 1, new updates after rolling are not visible
+  CloseDB();
+  cloud_env_options_.cookie_on_open = "1";
+  OpenDB();
+  std::string value;
+  ASSERT_OK(db_->Get(ReadOptions(), "Hello", &value));
+  ASSERT_EQ(value, "world");
```
Could you also verify that writes are visible with the new cookie? I think that's the scenario that doesn't work in the current version of the code.

seckcoder commented on cloud/cloud_env_impl.cc:
```diff
@@ -1840,6 +1843,45 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
+    const CloudManifestDelta& delta) {
+  std::string old_epoch = cloud_manifest_->GetCurrentEpoch();
+  const auto& fs = GetBaseEnv()->GetFileSystem();
+  auto io_st = CopyFile(fs.get(), ManifestFileWithEpoch(local_dbname, old_epoch),
+                     ManifestFileWithEpoch(local_dbname, delta.epoch),
+                     0 /* size */, true /* use_fsync */,
+                     nullptr /* io_tracer */, Temperature::kUnknown);
```
That's a good point. Thinking more about this.

seckcoder replied:
So if we make the local manifest filename just MANIFEST, then we don't need to do anything about local MANIFEST when rolling the epoch/CloudManifest. We just need to update local CloudManifest  with the new epoch, upload MANIFEST to s3 as MANIFEST-new_epoch and upload  CLOUDMANIFEST-cookie. That seems to be able to indeed solve the issue.

igorcanadi replied:
Exactly! I don't really know why we added -suffix to the local MANIFEST filename, I can't think of a single good reason to do it.

igorcanadi replied:
1. You can't write to the existing `MANIFEST` file because you might get in the corrupted state if you crash mid-write. You can, however, write to `MANIFEST.tmp` and then do an atomic rename `MANIFEST.tmp -> MANIFEST`. I don't know why RocksDB decided to do this atomic switch through the `CURRENT` file instead of the atomic rename. Possibly to support filesystems that don't have a rename operation? Not a concern for us.
2. Today, yes, we write to the same `MANIFEST-epoch` file with the technique from (1).

seckcoder replied:
I see. since we are already writing to the same MANIFEST-epoch file, I guess there shouldn't be any issues with writing to `MANIFEST` file (with renaming of course).

seckcoder replied:
BTW, do we have concerns if the local `MANIFEST`/`CLOUDMANIFEST` file is not consistent with the one in the cloud? This can happen if we update local file correctly, but uploading to s3 fails.

To me, looks like this can be an issue. We always read from local storage first: https://github.com/rockset/rocksdb-cloud/blob/master/cloud/cloud_env_impl.cc#L172

Also, reopen db won't fix this issue?

seckcoder replied:
nvm. I think this is fine. We have the same issue today.

seckcoder replied:
I'll create a separate PR for this and gradually roll it out in prod.

seckcoder commented on cloud/cloud_env_impl.cc:
```diff
@@ -1840,6 +1843,45 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
+
+  cloud_manifest_->AddEpoch(delta.file_num, delta.epoch);
+
+  // Dump cloud_manifest into the CLOUDMANIFEST-new_cookie file
+  return writeCloudManifest(cloud_manifest_.get(),
+                            MakeCloudManifestFile(local_dbname, new_cookie));
```
About `CLOUDMANIFEST`, I think either is fine. There doesn't seem to be correctness issue with either approach.

seckcoder commented:
@igorcanadi PTAL . Cherry-picked the change of "new manifest on next update". Also addressed comments about:
1. [build RollNewEpoch on top of our newly-introduced functions](https://github.com/rockset/rocksdb-cloud/pull/167/commits/ab78f4ffba38462e62b5495a10cf5699bf71094e)
2. [upload the cloudmanifest file in the old format (CLOUDMANIFEST) even when the cookie is set](https://github.com/rockset/rocksdb-cloud/pull/167/commits/4ba3fc701d7d28f74c5005b336ea1fc240e9ff3e)

igorcanadi commented on cloud/cloud_env_impl.cc:
```diff
@@ -1796,47 +1803,113 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
+
+Status CloudEnvImpl::UploadLocalCloudManifest(const std::string& local_dbname,
+                                              const std::string& cookie) {
+   // upload the cloud manifest file corresponds to cookie (i.e., CLOUDMANIFEST-cookie)
```
This is a public function. We should fail if there is no DestBucket in case of the caller's error.

igorcanadi commented on cloud/cloud_env_impl.cc:
```diff
@@ -1796,47 +1803,113 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
+
+Status CloudEnvImpl::UploadLocalCloudManifestAndManifest(
+    const std::string& local_dbname, const std::string& cookie) {
+  assert(HasDestBucket());
```
Let's be nicer and return an error if there is no dest bucket. See also below.

igorcanadi reviewed (approved):
Epic, ship it.
